it/google-cloud-platform/src/main/java/org/apache/beam/it/gcp/TemplateTestBase.java

```diff
@@ -180,72 +188,115 @@ public void setUpBase() throws ExecutionException {
           "Both -DartifactBucket and -DstageBucket were not given. Storage Client will not be"
               + " created automatically.");
     }
-
     credentialsProvider = FixedCredentialsProvider.create(credentials);
-    pipelineLauncher = buildLauncher();
 
-    if (usingDirectRunner) {
-      // Using direct runner, not needed to stage.
-      return;
+    if (annotation != null) {
+      templateClass = annotation.value();
+      template = getTemplateAnnotation(annotation, templateClass);
+      if (template == null) {
+        return;
+      }
+      if (template.placeholderClass() != void.class) {
+        templateClass = template.placeholderClass();
+      }
+      pipelineLauncher = buildLauncher(templateClass, template);
+
+      if (usingDirectRunner) {
+        // Using direct runner, not needed to stage.
+        return;
+      }
+      specPath = getSpecPath(templateClass, template, "pom.xml");
+    } else {
+      TemplateIntegrationTest[] templateIntegrationTests = multiAnnotation.value();
+      String[] pomPaths = multiAnnotation.pomPaths();
+      if (pomPaths.length != templateIntegrationTests.length) {
+        LOG.warn(
+            "@MultiTemplateIntegrationTest fields value and pomPaths specified in {} have unequal"
+                + " lengths. Every template class in value should have its corresponding pom file"
+                + " in pomPath at the same index.",
+            getClass());
+        return;
+      }
+      for (int i = 0; i < templateIntegrationTests.length; i++) {
+        TemplateIntegrationTest templateIntegrationTest = templateIntegrationTests[i];
+        String pomPath = pomPaths[i];
+        templateClass = templateIntegrationTest.value();
+        template = getTemplateAnnotation(templateIntegrationTest, templateClass);
+        if (template == null) {
+          return;
+        }
+        if (template.placeholderClass() != void.class) {
+          templateClass = template.placeholderClass();
+        }
+        multiTemplates.add(template);
+        multiTemplateClasses.add(templateClass);
+        pipelineLauncher = buildLauncher(templateClass, template);
```

Comment:
Do we need separate pipeline launchers per template when there are multiple templates?

Reply:
Added null check to do this only once. This earlier would be a redundant call re-initialising to the same value each time.

Reply:
In the launchTemplate method, when setupShutdownHook is enabled - a separate thread is created to which the pipelineLauncher object is passed. Will there be multi-threading issues if there are multiple templates and the same pipelineLauncher object is being used across multiple threads to cancel the job?

Reply:
I went through the beam code it looks like we should be safe. I could not find documentation explicitly saying it thread-safe though nor have i tested it.

Reply:
Cool, thank you.